src/quickunlock/QuickUnlockInterface.cpp

```diff
@@ -16,66 +16,63 @@
  */
 
 #include "QuickUnlockInterface.h"
+#include "PinUnlock.h"
+
 #include <QObject>
 
 #if defined(Q_OS_MACOS)
 #include "TouchID.h"
-#define QUICKUNLOCK_IMPLEMENTATION TouchID
 #elif defined(Q_CC_MSVC)
 #include "WindowsHello.h"
-#define QUICKUNLOCK_IMPLEMENTATION WindowsHello
 #elif defined(Q_OS_LINUX)
 #include "Polkit.h"
-#define QUICKUNLOCK_IMPLEMENTATION Polkit
-#else
-#define QUICKUNLOCK_IMPLEMENTATION NoQuickUnlock
 #endif
 
-QUICKUNLOCK_IMPLEMENTATION* quickUnlockInstance = {nullptr};
+QuickUnlockManager* g_quickUnlockManager = nullptr;
 
-QuickUnlockInterface* getQuickUnlock()
+QuickUnlockManager* getQuickUnlock()
 {
-    if (!quickUnlockInstance) {
-        quickUnlockInstance = new QUICKUNLOCK_IMPLEMENTATION();
+    if (!g_quickUnlockManager) {
+        g_quickUnlockManager = new QuickUnlockManager();
     }
-    return quickUnlockInstance;
-}
-
-bool NoQuickUnlock::isAvailable() const
-{
-    return false;
-}
-
-QString NoQuickUnlock::errorString() const
-{
-    return QObject::tr("No Quick Unlock provider is available");
+    return g_quickUnlockManager;
 }
 
-void NoQuickUnlock::reset()
+QuickUnlockManager::QuickUnlockManager()
 {
+    // Create the native interface based on the platform
+#if defined(Q_OS_MACOS)
+    m_nativeInterface.reset(new TouchID());
+#elif defined(Q_CC_MSVC)
+    m_nativeInterface.reset(new WindowsHello());
+#elif defined(Q_OS_LINUX)
+    m_nativeInterface.reset(new Polkit());
+#endif
+    // Always create the fallback interface
+    m_fallbackInterface.reset(new PinUnlock());
 }
 
-bool NoQuickUnlock::setKey(const QUuid& dbUuid, const QByteArray& key)
+QuickUnlockManager::~QuickUnlockManager()
 {
-    Q_UNUSED(dbUuid)
-    Q_UNUSED(key)
-    return false;
 }
 
-bool NoQuickUnlock::getKey(const QUuid& dbUuid, QByteArray& key)
+QSharedPointer<QuickUnlockInterface> QuickUnlockManager::interface() const
 {
-    Q_UNUSED(dbUuid)
-    Q_UNUSED(key)
-    return false;
+    if (isNativeAvailable()) {
+        return m_nativeInterface;
+    }
+    return m_fallbackInterface;
 }
 
-bool NoQuickUnlock::hasKey(const QUuid& dbUuid) const
+bool QuickUnlockManager::isNativeAvailable() const
 {
-    Q_UNUSED(dbUuid)
-    return false;
+    return m_nativeInterface && m_nativeInterface->isAvailable();
 }
 
-void NoQuickUnlock::reset(const QUuid& dbUuid)
+bool QuickUnlockManager::isRememberAvailable() const
 {
-    Q_UNUSED(dbUuid)
+    if (isNativeAvailable()) {
+        return m_nativeInterface->canRemember();
+    }
+    return m_fallbackInterface->canRemember();
 }
```

Comment:
Looks like this isn't used anywhere?

Reply:
Yah there was a purpose at some point but the settings shuffle made this unnecessary.

Reply:
This would be needed for an explicit reset if the key cannot be remembered. We should definitely do that, even if it may technically not be neccessary with the way these are implemented now. We should not make any assumptions here whether an explicit reset is necessary or not.

Reply:
I don't follow, there is nothing to reset if the quick unlock interface cannot find a stored credential set based on the public uuid.